src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,408 @@
+    // Calculate the number of nuclear zones
+    inline size_type num_nuclear_zones(AtomicMassNumber a);
+
+    // Calculate the nuclear radius
+    inline real_type calc_nuclear_radius(AtomicMassNumber a);
+
+    // Calculate components of nuclear zone properties
+    inline void
+    calc_zone_component(IsotopeView const& target, ComponentVec& components);
+
+    // Integrate the Woods-Saxon potential in [rmin, rmax]
+    inline real_type integrate_woods_saxon(real_type rmin,
+                                           real_type rmax,
+                                           real_type radius) const;
+
+    // Integrate the Gaussoan potential in [rmin, rmax]
+    inline real_type
+    integrate_gaussian(real_type rmin, real_type rmax, real_type radius) const;
+
+    // Integrate a nuclear potential by adaptive quadrature
+    inline real_type integrate_potential(real_type rmin,
+                                         real_type delta_r,
+                                         real_type integral1,
+                                         real_type ws_shift = 0) const;
+
+  private:
+    //// DATA ////
+
+    // Cascade model configurations and nuclear structure parameters
+    CascadeOptions const& options_;
+
+    real_type skin_depth_;
+    MevMass neutron_mass_;
+    MevMass proton_mass_;
+
+    CollectionBuilder<ZoneComponent> components_;
+    CollectionBuilder<NuclearZones, MemSpace::host, IsotopeId> zones_;
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type half() { return 0.5; }
+    static CELER_CONSTEXPR_FUNCTION real_type four_thirds_pi()
+    {
+        return 4 * constants::pi / real_type{3};
+    }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with cascade options and data.
+ */
+CELER_FUNCTION
+NuclearZoneBuilder::NuclearZoneBuilder(CascadeOptions const& options,
+                                       NeutronInelasticScalars const& scalars,
+                                       Data* data)
+    : options_(options)
+    , skin_depth_(0.611207 * options.radius_scale)
+    , neutron_mass_(scalars.neutron_mass)
+    , proton_mass_(scalars.proton_mass)
+    , components_(&data->components)
+    , zones_(&data->zones)
+{
+    CELER_EXPECT(options_);
+    CELER_EXPECT(data);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct nuclear zone data for a single isotope.
+ */
+void NuclearZoneBuilder::operator()(IsotopeView const& target)
+{
+    AtomicNumber a = target.atomic_mass_number();
+    size_type num_zones = this->num_nuclear_zones(a);
+
+    std::vector<ZoneComponent> comp(num_zones);
+    this->calc_zone_component(target, comp);
+
+    NuclearZones nucl_zones;
+    nucl_zones.num_zones = num_zones;
+    nucl_zones.zones = components_.insert_back(comp.begin(), comp.end());
```

Comment:
`num_zones` is redundant with `zones.size()`, so maybe that data should be removed?

Reply:
Okay, it is removed from the zone data (but keep the data structure as it was since more data may be added later).